contrib/cgi/src/cgi.c

```diff
@@ -0,0 +1,683 @@
+#include <config.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <glib.h>
+#include <glib/gstdio.h>
+#include <gio/gio.h>
+#include <json-glib/json-glib.h>
+#include <rauc-installer-generated.h>
+
+/*
+ * Bundle/status file locations
+ * It is assumed that these files do not survive a reboot.
+ */
+const gchar *BUNDLE_TARGET_LOCATION = "/tmp/bundle.raucb";
+const gchar *STATUS_FILE_LOCATION = "/tmp/rauc-status.json";
+
+/*
+ * HTTP request headers that should be saved during PUT and are then available
+ * in status method, lowercase, NULL-terminated
+ */
+const gchar *HTTP_HEADERS[] = {"upload_client_id", NULL};
+
+
+#define CGI_ERROR cgi_error_quark()
+
+static GQuark cgi_error_quark(void)
+{
+	return g_quark_from_static_string("cgi_error_quark");
+}
+
+typedef enum {
+	CGI_ERROR_FAILED,
+	CGI_ERROR_METHOD_NOT_ALLOWED,
+	CGI_ERROR_LENGTH_REQUIRED,
+	CGI_ERROR_BAD_REQUEST,
+	CGI_ERROR_SERVICE_UNAVAILABLE,
+} CGIError;
+
+/*
+ * Print HTTP headers and two new lines to start HTTP body.
+ */
+static void print_headers(const gchar* status, const gchar* content_type)
+{
+	g_print("Status: %s\n", status);
+	g_print("Content-type: %s\n", content_type);
+	g_print("\n");
+}
+
+/*
+ * Retrieve LastError property via rauc D-Bus service.
+ */
+static gchar* get_last_error(RInstaller *installer, GError **error)
+{
+	gchar* last_error = r_installer_dup_last_error(installer);
+	if (!last_error)
+		g_set_error(error, CGI_ERROR, CGI_ERROR_FAILED, "D-Bus error");
+
+	return last_error;
+}
+
+/*
+ * Retrieve Operation property via rauc D-Bus service.
+ */
+static gchar* get_operation(RInstaller *installer, GError **error)
+{
+	gchar* operation = r_installer_dup_operation(installer);
+	if (!operation)
+		g_set_error(error, CGI_ERROR, CGI_ERROR_FAILED, "D-Bus error");
+
+	return operation;
+}
+
+/*
+ * Retrieve ProgressUpdated property via rauc D-Bus service.
+ */
+static gboolean get_progress_updated(RInstaller *installer, gint *percentage, gchar **description, GError **error)
+{
+	gboolean res = FALSE;
+	GVariant* progress_updated;
+	GVariant* g_percentage;
+	GVariant* g_description;
+
+	*percentage = -1;
+
+	progress_updated = r_installer_get_progress(installer);
+	if (!progress_updated) {
+		g_set_error(error, CGI_ERROR, CGI_ERROR_FAILED, "D-Bus error");
+		goto out;
+	}
+
+	g_percentage = g_variant_get_child_value(progress_updated, 0);
+	*percentage = g_variant_get_int32(g_percentage);
+
+	g_description = g_variant_get_child_value(progress_updated, 1);
+	*description = g_strdup(g_variant_get_string(g_description, NULL));
+
+	res = TRUE;
+
+	g_variant_unref(g_percentage);
+	g_variant_unref(g_description);
+
+out:
+	return res;
+}
+
+/*
+ * Last installation was successfull if LastError D-Bus property is an empty
+ * string.
+ */
+static gboolean get_last_installation_success(RInstaller *installer, GError **error)
```

Comment:
See my note about gboolean return with GError ** argument above.

Reply:
Here your note from does actually apply, will fix it.